### PR TITLE
Support Python 3.6 to 3.10

### LINKED ISSUE #289: Requirement is pinned for pyicloud 0.9.7: click>=6.0,<=7.1.1
## The problem

- This library supports an integration in Home Assistant.
- The upcoming pip resolver will be more strict in resolving package version conflicts:
http://pyfound.blogspot.com/2020/03/new-pip-resolver-to-roll-out-this-year.html#what-will-change
- To avoid future issues it would be good if this library would change its pinning to allow any higher version and just set a minimum version.

Thanks for your support of Home Assistant!

## Environment

- pyiCloud release with the issue (`pip show pyicloud`): 0.9.7
- Last working pyiCloud release (if known): 
- Service causing this issue: 
- Python version (`python -V`): 3.7.7
- Operating environment (project deps/Docker/Windows/etc.):  Ubuntu 18.04

## Traceback/Error logs

```shell
(home-assistant) x:~/Dev/home-assistant/home-assistant$ pip install -r requirements_all.txt
...
(home-assistant) x:~/Dev/home-assistant/home-assistant$ pipdeptree --reverse --packages click
...
click==7.1.1
...
  - pyicloud==0.9.7 [requires: click>=6.0,<=7.1.1]
...
```

### LINKED ISSUE #290: Requirement is pinned for pyicloud 0.9.7: keyring>=8.0,<=9.3.1 and keyrings.alt>=1.0,<=3.2.0
## The problem

- This library supports an integration in Home Assistant.
- The upcoming pip resolver will be more strict in resolving package version conflicts:
http://pyfound.blogspot.com/2020/03/new-pip-resolver-to-roll-out-this-year.html#what-will-change
- To avoid future issues it would be good if this library would change its pinning to allow any higher version and just set a minimum version.

Thanks for your support of Home Assistant!

## Environment

- pyiCloud release with the issue (`pip show pyicloud`): 0.9.7
- Last working pyiCloud release (if known): 
- Service causing this issue: 
- Python version (`python -V`): 3.7.7
- Operating environment (project deps/Docker/Windows/etc.):  Ubuntu 18.04

## Traceback/Error logs

```shell
(home-assistant) x:~/Dev/home-assistant/home-assistant$ pip install -r requirements_all.txt
...
(home-assistant) x:~/Dev/home-assistant/home-assistant$ pipdeptree --reverse --packages keyring
...
keyring==21.2.0
  - pyicloud==0.9.7 [requires: keyring>=8.0,<=9.3.1]
(home-assistant) x:~/Dev/home-assistant/home-assistant$ pipdeptree --reverse --packages keyrings.alt
...
keyrings.alt==3.4.0
  - pyicloud==0.9.7 [requires: keyrings.alt>=1.0,<=3.2.0]
```

### PR DESCRIPTION
**Pull request by WhyNotHugo**
Rebase of #325. Looks like pushing to stale branches does not update the existing PR.

## Breaking change

- Python versions which are past their EOL are no long supported.-

## Proposed change

Drop support for Python versions which no longer have any upstream support. Improve forward compatibility up to Python 3.10.

This also should reduce dependency on older libraries, some of which are still pinned at old, insecure versions.

## Type of change

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New service (thank you!)
- [ ] New feature (which adds functionality to an existing service)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Documentation or code sample

## Example of code:

n/a

## Additional information

Fixes #289
Fixes #290
Closes #303
Closes #325

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] Tests have been added to verify that the new code works. (kinda, CI runs on newer pythons).

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated to README

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -143,15 +143,15 @@ Location
 Returns the device's last known location. The Find My iPhone app must have been installed and initialized.
 
 >>> api.iphone.location()
-{u'timeStamp': 1357753796553, u'locationFinished': True, u'longitude': -0.14189, u'positionType': u'GPS', u'locationType': None, u'latitude': 51.501364, u'isOld': False, u'horizontalAccuracy': 5.0}
+{u'timeStamp': 1357753796553, 'locationFinished': True, 'longitude': -0.14189, 'positionType': 'GPS', 'locationType': None, 'latitude': 51.501364, 'isOld': False, 'horizontalAccuracy': 5.0}
```
Comment (balloob) [inline README.rst:146]:
```suggestion
{'timeStamp': 1357753796553, 'locationFinished': True, 'longitude': -0.14189, 'positionType': 'GPS', 'locationType': None, 'latitude': 51.501364, 'isOld': False, 'horizontalAccuracy': 5.0}
```

File: README.rst [hunk 2]
```diff
@@ -216,18 +216,18 @@ You can access documents stored in your iCloud account by using the ``files`` pr
 
 >>> api.files.dir()
 [u'.do-not-delete',
```
Comment (balloob) [inline README.rst:218]:
```suggestion
['.do-not-delete',
```

File: README.rst [hunk 3]
```diff
@@ -216,18 +216,18 @@ You can access documents stored in your iCloud account by using the ``files`` pr
...
+ 'com~apple~system~spotlight'
 ]
 
 You can access children and their children's children using the filename as an index:
 
 >>> api.files['com~apple~Notes']
-<Folder: u'com~apple~Notes'>
+<Folder: 'com~apple~Notes'>
 >>> api.files['com~apple~Notes'].type
 u'folder'
```
Comment (balloob) [inline README.rst:232]:
```suggestion
'folder'
```

File: README.rst [hunk 4]
```diff
@@ -336,7 +336,7 @@ Note: Consider using ``shutil.copyfile`` or another buffered strategy for downlo
 Information about each version can be accessed through the ``versions`` property:
 
 >>> photo.versions.keys()
-[u'medium', u'original', u'thumb']
+[u'medium', 'original', 'thumb']
```
Comment (balloob) [inline README.rst:339]:
```suggestion
['medium', 'original', 'thumb']
```

File: pyicloud/base.py [hunk 5]
```diff
@@ -72,11 +71,11 @@ def request(self, method, url, **kwargs):  # pylint: disable=arguments-differ
         if self.service.password_filter not in request_logger.filters:
             request_logger.addFilter(self.service.password_filter)
 
-        request_logger.debug("%s %s %s" % (method, url, kwargs.get("data", "")))
+        request_logger.debug("{} {} {}".format(method, url, kwargs.get("data", "")))
```
Comment (balloob) [inline pyicloud/base.py:74]: Please use logging formatting such that the parameters are not stringified unless the log level is low enough to print it.

```suggestion
        request_logger.debug("%s %s %s", method, url, kwargs.get("data", ""))
```

File: README.rst [hunk 6]
```diff
@@ -204,7 +204,7 @@ You can access your iCloud contacts/address book through the ``contacts`` proper
 
 >>> for c in api.contacts.all():
 >>> print c.get('firstName'), c.get('phones')
-John [{u'field': u'+1 555-55-5555-5', u'label': u'MOBILE'}]
+John [{u'field': '+1 555-55-5555-5', 'label': 'MOBILE'}]
```
Comment (balloob) [inline README.rst:207]:
```suggestion
John [{'field': '+1 555-55-5555-5', 'label': 'MOBILE'}]
```

**Review by balloob — APPROVED** [review]
Looks great!

File: pyicloud/services/drive.py [hunk 7]
```diff
@@ -270,7 +285,7 @@ def open(self, **kwargs):
         return self.connection.get_file(self.data["docwsid"], **kwargs)
 
     def upload(self, file_object, **kwargs):
-        """"Upload a new file."""
+        """ "Upload a new file."""
```
Comment (MartinHjelmare) [inline pyicloud/services/drive.py:288]:
```suggestion
        """Upload a new file."""
```

File: pyicloud/services/ubiquity.py [hunk 8]
```diff
@@ -104,17 +103,11 @@ def get(self, name):
     def __getitem__(self, key):
         try:
             return self.get(key)
-        except IndexError:
-            raise KeyError("No child named %s exists" % key)
-
-    def __unicode__(self):
-        return self.name
+        except IndexError as i:
+            raise KeyError("No child named %s exists" % key) from i
```
Comment (MartinHjelmare) [inline pyicloud/services/ubiquity.py:107]:
```suggestion
            raise KeyError(f"No child named {key} exists") from i
```

File: tox.ini [hunk 9]
```diff
@@ -15,7 +17,6 @@ commands =
     {envbindir}/pytest
 
 [testenv:lint]
-basepython = python3
```
Comment (MartinHjelmare) [inline tox.ini:18]: Please keep this line.

File: pyicloud/base.py [hunk 10]
```diff
@@ -517,7 +516,8 @@ def trust_session(self):
 
         try:
             self.session.get(
-                "%s/2sv/trust" % self.AUTH_ENDPOINT, headers=headers,
+                "%s/2sv/trust" % self.AUTH_ENDPOINT,
```
Comment (MartinHjelmare) [inline pyicloud/base.py:519]:
```suggestion
                f"{self.AUTH_ENDPOINT}/2sv/trust",
```

File: requirements_test.txt [hunk 11]
```diff
@@ -1,6 +1,5 @@
 black==19.10b0
```
Comment (MartinHjelmare) [inline requirements_test.txt:1]: We'll need to bump black to support the latest Python versions.
```suggestion
black==22.1.0
```

**Comment by MartinHjelmare** [comment]
There are quite many pylint warnings for `consider-using-f-string` left. I think we should consider disabling that warning globally for now and solve that in a separate PR.

File: tox.ini (hunk 9 above)
Comment (WhyNotHugo) [reply tox.ini:18]: This really is a no-op nowadays, since the only supported Python is v3.

File: requirements_test.txt (hunk 11 above)
Comment (WhyNotHugo) [reply requirements_test.txt:1]: This is only necessary to support newer Python syntax, but given that we're keeping Python3.6 support, the previous version should be fine too.

File: tests/test_account.py [hunk 12]
```diff
@@ -6,7 +6,7 @@
 
 
 class AccountServiceTest(TestCase):
-    """"Account service tests"""
+    """ "Account service tests"""
```
Comment (MartinHjelmare) [inline tests/test_account.py:9]: Extra quote leftover.

File: tests/test_drive.py [hunk 13]
```diff
@@ -6,7 +6,7 @@
 
 # pylint: disable=pointless-statement
 class DriveServiceTest(TestCase):
-    """"Drive service tests"""
+    """ "Drive service tests"""
```
Comment (MartinHjelmare) [inline tests/test_drive.py:9]: As above.

File: tests/test_findmyiphone.py [hunk 14]
```diff
@@ -5,7 +5,7 @@
 
 
 class FindMyiPhoneServiceTest(TestCase):
-    """"Find My iPhone service tests"""
+    """ "Find My iPhone service tests"""
```
Comment (MartinHjelmare) [inline tests/test_findmyiphone.py:8]: As above.

**Review by MartinHjelmare — COMMENTED** [review]
Just noting some things for a future clean up.

**Comment by balloob** [comment]
Fixed final comments in 8c7ba2a

**Pull request merged**